pipeline/data_store.py

```diff
@@ -1172,51 +1105,41 @@ def get_subtraction(self, provenance=None, session=None):
-                provenances = []
-            if len(provenances) > 1:
-                raise ValueError(f'More than one "{process_name}" provenance found!')
-            if len(provenances) > 0:
-                # a mismatch of provenance and cached subtraction image:
-                if self.sub_image.provenance.id != provenances[0].id:
-                    self.sub_image = None  # this must be an old subtraction image, need to get a new one
+        # TODO: do we need to test the subtraction Provenance has upstreams consistent with upstream provenances?
 
         # not in memory, look for it on the DB
         if self.sub_image is None:
             with SmartSession(session, self.session) as session:
                 image = self.get_image(session=session)
                 ref = self.get_reference(session=session)
 
-                # this happens when the subtraction is required as an upstream for another process (but isn't in memory)
-                if provenance is None:  # check if in upstream_provs/database
-                    provenance = self._get_provenance_for_an_upstream(process_name, session=session)
-
-                if provenance is not None:  # if None, it means we can't find it on the DB
-                    aliased_table = sa.orm.aliased(image_upstreams_association_table)
-                    self.sub_image = session.scalars(
-                        sa.select(Image).join(
-                            image_upstreams_association_table,
-                            sa.and_(
-                                image_upstreams_association_table.c.upstream_id == ref.image_id,
-                                image_upstreams_association_table.c.downstream_id == Image.id,
-                            )
-                        ).join(
-                            aliased_table,
-                            sa.and_(
-                                aliased_table.c.upstream_id == image.id,
-                                aliased_table.c.downstream_id == Image.id,
-                            )
-                        ).where(Image.provenance.has(id=provenance.id))
-                    ).first()
+                aliased_table = sa.orm.aliased(image_upstreams_association_table)
+                self.sub_image = session.scalars(
+                    sa.select(Image).join(
+                        image_upstreams_association_table,
+                        sa.and_(
+                            image_upstreams_association_table.c.upstream_id == ref.image_id,
+                            image_upstreams_association_table.c.downstream_id == Image.id,
+                        )
+                    ).join(
+                        aliased_table,
+                        sa.and_(
+                            aliased_table.c.upstream_id == image.id,
+                            aliased_table.c.downstream_id == Image.id,
+                        )
+                    ).where(Image.provenance.has(id=provenance.id))
+                ).first()
```

Comment:
Here was the big query I mentioned in the meeting... I intend to come back to this and look into it some more, its far more complex than anything I have had to do!

Reply:
I can barely remember how I got it to work, too...

Reply:
`curl --data "Please write me sqlalchemy code to join together all my image upstreams correctly" https://chatgpt.org`